src/elements.jl

```diff
@@ -0,0 +1,144 @@
+
+"""Return the number of valence electrons"""
+n_elec_valence(el::AbstractElement) = charge_ionic(el)
+
+"""Return the number of core electrons"""
+n_elec_core(el::AbstractElement) = charge_nuclear(el) - charge_ionic(el)
+
+function local_potential_fourier(el::AbstractElement, q::AbstractVector)
+    local_potential_fourier(el, norm(q))
+end
+
+function local_potential_real(el::AbstractElement, q::AbstractVector)
+    local_potential_real(el, norm(q))
+end
+
+
+"""
+Element interacting with electrons via a bare Coulomb potential
+(for all-electron calculations)
+"""
+struct ElementAllElectron <: AbstractElement
+    Z::Int  # Nuclear charge
+    symbol  # Element symbol
+end
+charge_ionic(el::ElementAllElectron) = el.Z
+
+function ElementAllElectron(key)
+    ElementAllElectron(periodic_table[key].number, Symbol(periodic_table[key].symbol))
+end
+
+
+"""Radial local potential, in Fourier space: V(q) = int_{R^3} V(x) e^{-iqx} dx."""
+function local_potential_fourier(el::ElementAllElectron, q::T) where {T <: Real}
+    q == 0 && return zero(T)  # Compensating charge background
+    # General atom => Use default Coulomb potential
+    # We use int_R^3 1/r e^{-i q⋅x} = 4π / |q|^2
+    return -4T(π) * el.Z / q^2
+end
+
+"""Radial local potential, in real space."""
+local_potential_real(el::ElementAllElectron, r::Real) = -el.Z / r
+
+
+"""
+Element interacting with electrons via a pseudopotential model
+"""
+struct ElementPsp <: AbstractElement
+    Z::Int  # Nuclear charge
+    symbol  # Element symbol
+    psp     # Pseudopotential data structure
+end
+ElementPsp(key; psp) = ElementPsp(periodic_table[key].number,
+                                  Symbol(periodic_table[key].symbol), psp)
+charge_ionic(el::ElementPsp) = el.psp.Zion
+
+function local_potential_fourier(el::ElementPsp, q::T) where {T <: Real}
+    q == 0 && return zero(T)  # Compensating charge background
+    eval_psp_local_fourier(el.psp, q)
+end
+
+function local_potential_real(el::ElementPsp, r::Real)
+    # Use local part of pseudopotential defined in Element object
+    return eval_psp_local_real(el.psp, r)
+end
+
+
+"""
+Element where the interaction with electrons is modelled
+as in [CohenBergstresser1966] (DOI 10.1103/PhysRev.141.789)
+
+Only the homonuclear lattices of the diamond structure
+are implemented (i.e. Si, Ge, Sn).
+"""
+struct ElementCohenBergstresser <: AbstractElement
+    Z::Int  # Nuclear charge
+    symbol  # Element symbol
+    V_sym   # Map |G|^2 (in units of (2π / lattice_constant)^2) to form factors
+    lattice_constant  # Lattice constant (in Bohr) which is assumed
+end
+charge_ionic(el::ElementCohenBergstresser) = 2
+
+function ElementCohenBergstresser(key; lattice_constant=nothing)
+    element = periodic_table[key]
+
+    # Form factors from Cohen-Bergstresser paper Table 2, converted to Bohr
+    # Lattice constants from Table 1, converted to Bohr
+    V_sym_paper = Dict{Int64,Float64}()
+    if element.symbol == "Si"
```

Comment:
Put the data (vsyms and lattice constants) in a table to avoid code duplication?

Reply:
Not sure what you mean exactly. Maybe what I do in the Cohen-Bergstresser example?

Reply:
pot_elements = Dict("Si" => [-.21, .04, .08]); lattice_constants = Dict("Si" => 5.66); etc

Reply:
Nah, because it might not be just those three G^2 values, so that is actually too restrictive. And if you make Dict("Si" => Dict(4 => -.21, 5 => .04, 7 => .08]); lattice_constants = Dict("Si" => 5.66)
that becomes convoluted too.

Reply:
It is restrictive, that's kind of the point : we are not going to generalize cohen-bergstresser, it's a simple demo example. I'm just saying, you get the same result with less copy/paste. And if we want to generalize later on it's simpler to have to not do changes three times.

Reply:
Ok, I'll use a dict, but I'd like to make the units explicit. Makes it still a bit better, though.

Reply:
Sorry to be a pain: it's fine like this, we don't particularly care about this code anyway, I just have a visceral hate of this kind of copy-pasted code (essentially because I always screw it up)